keps/sig-api-machinery/20191001-oasv3.md

```diff
@@ -0,0 +1,949 @@
+            "type": "integer",
+            "format": "int64"
+          },
+          "resourceVersion": {
+            "description": "String that identifies the server's internal version of this object that can be used by clients to determine when objects have changed. Value must be treated as opaque by clients and passed unmodified back to the server. Populated by the system. Read-only. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency",
+            "type": "string"
+          },
+          "selfLink": {
+            "description": "selfLink is a URL representing this object. Populated by the system. Read-only.\n\nDEPRECATED Kubernetes will stop propagating this field in 1.20 release and the field is planned to be removed in 1.21 release.",
+            "type": "string"
+          }
+        }
+      },
+      "meta.v1.Patch": {
+        "description": "Patch is provided to give a concrete name and type to the Kubernetes PATCH request body.",
+        "type": "object"
+      }
+    },
+    "parameters": {
+      "query.pretty.5ca316":  {
+        "name": "pretty",
+        "in": "query",
+        "description": "If 'true', then the output is pretty printed.",
+        "schema": {
+          "type": "string"
+        }
+      }
+    }
+  }
+}
+```
+
+### New names for objects
+
+The old spec uses the full path to a package as a name for an object, for example `ConfigMap` is defined as `io.k8s.api.core.v1.ConfigMap`.
+This not only doesn't add any value but also reveals the internal path to a type. 
+
+Instead the new names will contain information about a group, version and kind(`GVK`). `ConfigMap` in the new format would become `core.v1.ConfigMap`.
+
+### Reusable parameters
+
+In order to make the spec smaller and more clearer common parameters for operations will be kept under `components/parameters`. 
+The [OASv3 spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#parameter-object) says that a unique parameter is defined by a combination of a name and location. 
+Thus to make a parameter unique across all operations its name will be derived as `name.location.SHA1(spec3.Parameter.MarshalJSON()) | hex.EncodeToString() | hexEncodedString[1:5]`.
```

Comment:
Would this be for parameter types like `CreateOptions`? For that I was expecting an identifier like `meta.v1.CreateOptions`.

It would seem better to share parameter definitions because we know they're the same (e.g. they share the same go type), not because they happen to have the same same json representation. E.g. have a single parameter definition for `CreateOptions` because we know they're all the same type, but not for two paramters that happen to look the same but have different purposes.

Reply:
this will be primarily for `query nad path` parameters for operations, that contain simple types like `string, boolean`. 

types like `CreateOptions` are now defined in a [RequestBody](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#requestBodyObject) field and will contain `$ref` identifier pointing to a definition, for example:

https://github.com/kubernetes/enhancements/pull/1263/files#diff-e3ca43e3e2f48d58ccbbb9f49671c017R712

https://github.com/kubernetes/enhancements/pull/1263/files#diff-e3ca43e3e2f48d58ccbbb9f49671c017R306

Reply:
The options are both on the body and in query params. For example, `dryRun`, `gracePeriodSeconds`, and all the `DeleteOptions` are also listed as query params: https://github.com/kubernetes/enhancements/pull/1263/files#diff-e3ca43e3e2f48d58ccbbb9f49671c017R720.

Since we know the type these parameters are defined with, would it be reasonable to have those represented in OpenAPI with something like this?

```
...
      "parameters": [
          {
            "name": "dryRun",
            "in": "query"
            "schema": {
              "$ref": "#/components/parameters/meta.v1.DeleteOptions/dryRun"
            }
          },
       ]
...
```

Reply:
not all `dryRun` parameters are equal, here they have different types:

`dryRun` from an operation:
```
{
  "description": "When present, indicates that modifications should not be persisted. An invalid or unrecognized dryRun directive will result in an error response and no further processing of the request. Valid values are: - All: all dry run stages will be processed",
  "in": "query",
  "name": "dryRun",
  "type": "string",
  "uniqueItems": true
}
```

`dryRun` from `meta.v1.DeleteOptions`:
```
{
  "description": "When present, indicates that modifications should not be persisted. An invalid or unrecognized dryRun directive will result in an error response and no further processing of the request. Valid values are: - All: all dry run stages will be processed",
  "type": "array",
  "items": {
    "type": "string"
  }
}
```

sometimes `gracePeriodSeconds` differ in `format`:

`gracePeriodSeconds` from `meta.v1.DeleteOptions`:
```
"gracePeriodSeconds": {
  "description": "The duration in seconds before the object should be deleted. Value must be non-negative integer. The value zero indicates delete immediately. If this value is nil, the default grace period for the specified type will be used. Defaults to a per object value if not specified. zero means delete immediately.",
  "type": "integer",
  "format": "int64"
}
```

`gracePeriodSeconds` from  an operation:
```
{
  "description": "The duration in seconds before the object should be deleted. Value must be non-negative integer. The value zero indicates delete immediately. If this value is nil, the default grace period for the specified type will be used. Defaults to a per object value if not specified. zero means delete immediately.",
  "in": "query",
  "name": "gracePeriodSeconds",
  "type": "integer",
  "uniqueItems": true
}
```

Sometimes the parameters differ only in one field, like in this example, only `description` is different:
```
{
  "name": "name",
  "in": "path",
  "description": "name of the ComponentStatus",
  "required": true,
  "schema": {
    "type": "string"
  }
}
```

```
{
  "name": "name",
  "in": "path",
  "description": "name of the ConfigMap",
  "required": true,
  "schema": {
    "type": "string"
  }
}
```

Reply:
even if we could tell they are the same then I think we would like to change `meta.v1.DeleteOptions` to reference the common parameter:
```
"meta.v1.DeleteOptions": {
  ...
  "properties": {
    "dryRun": {
       "$ref": "#/components/parameters/meta.v1.DeleteOptions/dryRun"
    }
    …
}
```

but it looks strange to reference `components/parameters` from `components/schemas`.

if this idea makes the spec hard to read then I am fine with abandoning it.

Reply:
Interesting. I hadn't considered restructuring `meta.v1.DeleteOptions` to reference the params. It might make more sense to just put the options params in `components/schemas` (instead of `components/parameters`) if we did that. But when I first saw it, my initial thought was also that it might be inconvenient to read. Even if we don't take it that far, having the param schemas consolidated could be an improvement over what we have today.

Thanks for digging in on this.

Reply:
currently `metav1.DeleteOptions` is supported both in HTTP body and url query parameter, and the openapi spec is correctly reflecting that. A couple of bugs:

- `dryRun` from an operation query parameter shouldn't be of `string` type. The in memory structure is a string slice and I believe our query parser supports array. 
- `gracePeriodSeconds` from an operation lacks format: we can [add format detail](https://github.com/kubernetes/kubernetes/blob/3317805652d145c4003f5c534c7d87fb5cd87d50/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L996-L1003) when we generate the spec

The query parameter and the body parameter should be consistent since they are both generated from the go type. I don't know if OAS v3 supports referring query parameters like https://github.com/kubernetes/enhancements/pull/1263/files#r330702453 suggested though

Reply:
I think that they point to the same members of the same struct.

It seems like both [the typed client](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_type.go#L465) and [the dynamic client](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/dynamic/simple.go#L218) always use `metav1.DeleteOptions` as a body parameter. Moreover, all the other parameters are passed as query parameters, why `metav1.DeleteOptions` are special?

`metav1.DeleteOptions` as a query parameter doesn't have [preconditions](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L474) field, this might be due to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L970

Reply:
Knowing that they should point to the same structure I think we could just use the scheme. For example:

```
{
  "name": "DeleteOptions",
  "in": “query”,
  "schema": {
    "$ref": "#/components/schemas/meta.v1.DeleteOptions
  }
 …
}
```

we could use `style` and `explode` keywords to control the serialization of the parameter - https://swagger.io/docs/specification/serialization/

Reply:
> It seems like both the typed client and the dynamic client always use metav1.DeleteOptions as a body parameter. Moreover, all the other parameters are passed as query parameters, why metav1.DeleteOptions are special?

it's because we can't express nested structs in the query parameters. all fields are flattened in queries.

Reply:
> we could use style and explode keywords to control the serialization of the parameter - https://swagger.io/docs/specification/serialization/

👍🏻the default setting i.e `style=form,explode=true` is working for us. but the specification doesnt seem to define how to deal with nested structures from the schema object.

Reply:
> why metav1.DeleteOptions are special?

also see: https://github.com/kubernetes/kubernetes/issues/34856